### PR TITLE
fix: paste before copy resulted in value being undefined

### LINKED ISSUE #8941: [Bug] TypeError: Cannot read properties of undefined (reading 'length')
Sentry Issue: [APPSMITH-1AC](https://sentry.io/organizations/appsmith/issues/2758259928/?referrer=github_integration)

```
TypeError: Cannot read properties of undefined (reading 'length')
  at isCopiedModalWidget (sagas/WidgetOperationUtils.ts:319:26)
  at call (sagas/WidgetOperationSagas.tsx:751:14)
...
(32 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by aswathkk**
## Description

If we paste before copying anything, `CopiedWidgetGroups` is undefined which results in some errors down the line.

Fixes #8941 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Manually:
1. Create an application in Incognito mode (just to make sure your copy history in indexed storage is empty)
2. Press `Cmd` + `V`

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:white_circle: Total coverage has not changed</summary>

    // Code coverage diff between base branch:release and head branch: fix/paste-before-copy 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | app/client/src/sagas/WidgetOperationSagas.tsx | 53.95 **(-0.02)** | 42.16 **(0.08)** | 53.19 **(0)** | 55.83 **(0.12)**
 :red_circle: | app/client/src/utils/storage.ts | 33.96 **(-0.22)** | 15.79 **(-0.88)** | 9.38 **(0)** | 26.05 **(-0.22)**</details>

### CODE REVIEW

File: app/client/src/sagas/WidgetOperationSagas.tsx [hunk 1]
```diff
@@ -707,6 +707,7 @@ export function calculateNewWidgetPosition(
  */
 function* pasteWidgetSaga(action: ReduxAction<{ groupWidgets: boolean }>) {
   let copiedWidgetGroups: CopiedWidgetGroup[] = yield getCopiedWidgets();
+  if (!copiedWidgetGroups) return;
```
Comment (marks0351) [inline app/client/src/sagas/WidgetOperationSagas.tsx:710]: is there a way to add jest test case for this by mocking getCopiedWidgets?

**Review by marks0351 — APPROVED** [review]

File: app/client/src/sagas/WidgetOperationSagas.tsx (hunk 1 above)
Comment (aswathkk) [reply app/client/src/sagas/WidgetOperationSagas.tsx:710]: @marks0351 
Do we need that? This is an implementation detail. Right?
This should've been caught if we had proper type definition for `getCopiedWidgets`.

Comment (marks0351) [reply app/client/src/sagas/WidgetOperationSagas.tsx:710]: yea can we have the proper type definition for `getCopiedWidgets` then.

Comment (aswathkk) [reply app/client/src/sagas/WidgetOperationSagas.tsx:710]: @marks0351 I was not able to make it work as expected. `copiedWidgetGroups` always being `any`.
Have you tried [typed-redux-saga](https://github.com/agiledigital/typed-redux-saga)? I think this library can provide better typescript support with redux-saga.

Comment (marks0351) [reply app/client/src/sagas/WidgetOperationSagas.tsx:710]: yea type definition along with sagas are difficult but i think copiedWidgetGroups is type defined but getCopiedWidgets is not, when STORAGE_KEYS.COPIED_WIDGET is null or empty array it doesn't seem to return anything. getCopiedWidgets can be type defined to an array and make sure it returns an array unless it goes the error path.

**Review by marks0351 — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@aswathkk Verified this PR in incognito mode and normal chrome browser by clearing all history and cache - now  (reading 'length') error is not occurring in chrome browser

But this still happens in safari browser after clearing cache

Screenshot 2021-11-04 at 10 18 54 AM

**Comment by aswathkk** [comment]
@YogeshJayaseelan I could see that you were using release.app.appsmith.com rather than using the DP.

**Comment by vivekverma2312** [comment]
@aswathkk Tested the DP on chrome,safari and firefox. Working fine in these browsers.

**Pull request merged**